src/Agent.Plugins/PipelineArtifact/FilePathUNCServer.cs

```diff
@@ -0,0 +1,95 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Agent.Sdk;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.VisualStudio.Services.WebApi;
+using Microsoft.VisualStudio.Services.Agent.Util;
+using Microsoft.VisualStudio.Services.Content.Common.Tracing;
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    public class FilePathUNCServer
+    {
+        internal async Task UploadAsync(
+            AgentTaskPluginExecutionContext context,
+            Guid projectId,
+            int buildId,
+            string artifactName,
+            string targetPath,
+            string fileSharePath,
+            CancellationToken token)
+        {
+            string artifactPath = Path.Join(fileSharePath, artifactName);
+
+            // create the artifact. at this point, mkdir already succeeded so the path is good.
+            // the artifact should get cleaned up during retention even if the copy fails in the
+            // middle
+            Directory.CreateDirectory(artifactPath);
+
+            // 2) associate the pipeline artifact with an build artifact
+            VssConnection connection = context.VssConnection;
+            BuildServer buildServer = new BuildServer(connection);
+            var propertiesDictionary = new Dictionary<string, string>
+            {
+                { FileShareArtifactUploadEventProperties.ArtifactName, artifactName },
+                { FileShareArtifactUploadEventProperties.ArtifactType, PipelineArtifactConstants.FileShareArtifact },
+                { FileShareArtifactUploadEventProperties.ArtifactLocation, fileSharePath }
+            };
+
+            var artifact = await buildServer.AssociateArtifactAsync(projectId, buildId, artifactName, ArtifactResourceTypes.FilePath, fileSharePath, propertiesDictionary, token);
+            var parallel = context.GetInput(FileShareArtifactUploadEventProperties.Parallel, required: false);
+            var parallelCount = parallel == "true" ? GetParallelCount(context, context.GetInput(FileShareArtifactUploadEventProperties.ParallelCount, required: false)) : 1;
+
+            // To copy all the files in one directory to another directory.
+            // Get the files in the source folder. (To recursively iterate through
+            // all subfolders under the current directory, see
+            // "How to: Iterate Through a Directory Tree.")
```

Comment:
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/file-system/how-to-iterate-through-a-directory-tree

"NTFS file systems can contain reparse points in the form of junction points, symbolic links, and hard links. The .NET Framework methods such as GetFiles and GetDirectories will not return any subdirectories under a reparse point. This behavior guards against the risk of entering into an infinite loop when two reparse points refer to each other."

So this new way won't work with subdirectories under a reparse point, which I think is fine? @willsmythe, can you comment on this?

Reply:
"NTFS file systems can contain reparse points in the form of junction points, symbolic links, and hard links. The .NET Framework methods such as GetFiles and GetDirectories will not return any subdirectories under a reparse point. This behavior guards against the risk of entering into an infinite loop when two reparse points refer to each other. "

Reply:
Ok, so currently If I'm downloading a list of files/directories from a symbolic directory say c://symbolicFolder to c://destination folder, this scenario will work in the current build artifact task. For both GetFiles and GetDirectories functions, this behavior is not supported, the reason being the description above. @willsmythe Should this scenario be covered? Or it is ok to not support symbolic links/reparse points. I'm not sure if build artifact is handling the scenario where there is a loop situation in the file system.

Reply:
I don't think we can/should go out of our way to support behavior that .NET had reasons not to support. 

It would be good to know exactly how the current Build Artifact tasks work so we can accurately explain the differences between these tasks and the Pipeline Artifact tasks. Can you find this out?

Reply:
To be super super clear, the behavior should be what the behavior already is - otherwise we will break people.